### PR TITLE
KEYCLOAK-18562 JPA clients no-downtime store

### PR DESCRIPTION
**Pull request by vramik**

### CODE REVIEW

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientModelCriteriaBuilder.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                        cb.equal(root.get(modelField.getName()), value[0])
+                    );
+                } else if (modelField.equals(SearchableFields.SCOPE_MAPPING_ROLE)) {
+                    validateValue(value, 1, modelField, op);
+
+                    return new JpaClientModelCriteriaBuilder((cb, root) -> 
+                        cb.isTrue(cb.function("@>",
+                            Boolean.TYPE,
+                            cb.function("->", JsonbType.class, root.get("metadata"), cb.literal("fScopeMappings")),
+                            cb.literal("\"" + value[0] + "\"")))
```
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientModelCriteriaBuilder.java:72]: (SQL injection) Can `value[0]` contain `"`?

File: model/map-jpa/src/main/resources/META-INF/persistence.xml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<persistence version="2.1" xmlns="http://xmlns.jcp.org/xml/ns/persistence" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/persistence http://xmlns.jcp.org/xml/ns/persistence/persistence_2_1.xsd">
+  <persistence-unit name="keycloak-client-store" transaction-type="RESOURCE_LOCAL">
```
Comment (hmlnarik) [inline model/map-jpa/src/main/resources/META-INF/persistence.xml:3]: Is `transaction-type` necessary?

File: model/map-jpa/src/main/resources/META-INF/jpa-client-changelog.xml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                protocol varchar(36) generated always as (metadata->>'fProtocol') stored,
+                enabled boolean generated always as ((metadata->>'fEnabled')::boolean) stored,
+                metadata jsonb
+            );
+
+            create index client_entityVersion on client(entityVersion);
+            create index client_realmId on client(realmId);
+            create index client_clientId on client(clientId);
+            create index client_scopeMappings on client using gin ((metadata->'fScopeMappings') jsonb_path_ops);
+            create index client_clientId_gin_trgm_idx on client using gin (clientId gin_trgm_ops);
```
Comment (hmlnarik) [inline model/map-jpa/src/main/resources/META-INF/jpa-client-changelog.xml:41]: If not necessary, please do not introduce trigram index

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+@TypeDefs({@TypeDef(name = "jsonb", typeClass = JsonbType.class)})
+public class JpaClientEntity extends AbstractClientEntity implements Serializable {
+
+    @Id
+    @Column
+    private UUID id;
+
+    @Type(type = "jsonb")
+    @Column(columnDefinition = "jsonb")
+    private final JpaClientMetadata metadata = new JpaClientMetadata();
```
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java:60]: Perhaps a shared instance could be used here rather than new instance with every `JpaClientEntity`?
```suggestion
    private final JpaClientMetadata metadata = JpaClientMetadata.INSTANCE;
```
Or maybe even `null`?

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Great job @vramik!
Could you please check the questions inline?
Also please move the `org.keycloak.models.map.jpa` package to `org.keycloak.models.map.storage.jpa` to be consistent with `chm` and Hot Rod (fixed in latest main)

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapKeycloakTransaction.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+// TODO find out the flush and clear https://thorben-janssen.com/5-common-hibernate-mistakes-that-cause-dozens-of-unexpected-queries/#Remove_Child_Entities_With_a_Bulk_Operation
+//        em.flush();
+//        em.clear();
+
+        return em.createQuery(deleteQuery).executeUpdate();
+    }
+
+    /**
+     * @return all clients with lower entityVersion field than {@code JpaClientMapStorage.SUPPORED_VERSION} as a stream.
+     */
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapKeycloakTransaction.java:159]: Implementation uses `entityVersion < JpaClientMapStorage.SUPPORED_VERSION`. I believe it's equivalent, isn't it?

[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapKeycloakTransaction.java:159]: Let `SUPPORED_VERSION = 3` and `entityVersion = 3`.
- `entityVersion <= JpaClientMapStorage.SUPPORED_VERSION + 1` is `3 <= 3 + 1` which is `true`
- `entityVersion < JpaClientMapStorage.SUPPORED_VERSION` is `3 < 3` which is `false`

[the remark this answers is not in the document]
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapKeycloakTransaction.java:159]: Don't we need only entities which are not migrated yet? When `SUPPORTED_VERSION = 3` and `entityVersion = 3` it means entity is up to date, doesn't t? I am  probably missing something here.

[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapKeycloakTransaction.java:159]: Perhaps it's me - I'm missing the purpose of this method. It seems it's used nowhere in this PR. What is the reason for introducing it?

[the remark this answers is not in the document]
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapKeycloakTransaction.java:159]: The idea behind it was to provide a way to support https://issues.redhat.com/browse/KEYCLOAK-18719. When there is upgrade between two incompatible versions and administrator will need to obtain all clients which are not migrated yet to perform migration on them. Does it make sense to have it here?

[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapKeycloakTransaction.java:159]: I think there needs to be an amendment. The histogram (KEYCLOAK-18719) is about getting a summary information: entityVersion -> count of rows with that entity version.
If there is a need to get the list of IDs of old entity version rows, then there should be no other columns selected, only that ID. The very structure of the table might change which means that one of the columns that's being selected here might not exist yet / anymore.

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java [hunk 6]
```diff
@@ -0,0 +1,622 @@
...
+
+    @Generated(GenerationTime.NEVER)
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java:64]: `@Column(insertable = false, updatable = false)` is necessary for telling hibernate that the field is not insertable and updatable, without it hibernate tries to update the field.
`@Basic(fetch = FetchType.LAZY)` is also necessary to mark field as lazy loaded, without it, the field would be loaded every time
`@Generated(GenerationTime.NEVER)` seems unnecessary, during development I had to add it (I don't remember exactly the reason), but when I test it now without it tests passes. Do we want to remove it?

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapKeycloakTransaction.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapKeycloakTransaction.java:159]: - it should be added new method something like `public Integer getCountByEntityVersion(Integer version)`
- `readNotMigratedClientStream` could remain there, but we should change it to select only id.

is that right?

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapStorageProviderFactory.java [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+                    String schema = config.get("schema");
+                    if (schema != null) {
+                        properties.put(JpaUtils.HIBERNATE_DEFAULT_SCHEMA, schema);
+                    }
+
+                    properties.put("hibernate.show_sql", config.getBoolean("showSql", false));
+                    properties.put("hibernate.format_sql", config.getBoolean("formatSql", true));
+                    properties.put("hibernate.dialect", config.get("driverDialect"));
+
+                    properties.put(AvailableSettings.ISOLATION, String.valueOf(Connection.TRANSACTION_REPEATABLE_READ));
```
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapStorageProviderFactory.java:146]: @hmlnarik I wanted to highlight this change, with default isolation level, concurrent tests fails, but I believe we cannot set the level to `TRANSACTION_REPEATABLE_READ`, right?

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java (hunk 4 above)
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java:60]: Unfortunately it cannot. With both shared instance and `null` there are `NPE`s.

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientModelCriteriaBuilder.java (hunk 1 above)
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientModelCriteriaBuilder.java:72]: Right now it cannot, because we call searchable field `SCOPE_MAPPING_ROLE` only from `MapClientProvider.preRemove`: https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/client/MapClientProvider.java#L354

And there is `RoleModel.getId()` used as value.

But I tried to test it to be on safe site and it seems it should be ok (if I understand it correctly). I tried to use `.compare(SearchableFields.SCOPE_MAPPING_ROLE, Operator.EQ, "\"::jsonb=true;drop schema public cascade;")` and the result was

```
Caused by: org.postgresql.util.PSQLException: ERROR: invalid input syntax for type json
  Detail: Token ""::jsonb=true;drop schema public cascade;" is invalid.
  Where: JSON data, line 1: "::jsonb=true;drop schema public cascade;
```

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapKeycloakTransaction.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapKeycloakTransaction.java:159]: I think that for this PR, it should be removed and KEYCLOAK-18719 would be handled separately once needed for migration.

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java (hunk 4 above)
Comment (hmlnarik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java:60]: Why is there NPE with a shared instance?

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientModelCriteriaBuilder.java (hunk 1 above)
Comment (hmlnarik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientModelCriteriaBuilder.java:72]: Then I don't understand there is the `"\""` in the `literal`. Should it be without?

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java:64]: IIRC, if the field has no `@Column` / `@Basic` annotation, it should not be handled by JPA at all. Do we want these fields to be handled by JPA?

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/liquibase/JpaClientLiquibaseUpdaterProviderFactory.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    public void postInit(KeycloakSessionFactory factory) {
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public String getId() {
+        return "client-updater";
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/liquibase/JpaClientLiquibaseUpdaterProviderFactory.java:46]: Perhaps would be also good to have more specific name, like `map-jpa-client-updater`. However I wonder if we even need a separate provider for this at all. Is map JPA usable without this specific updater?

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapStorageProviderFactory.java (hunk 7 above)
Comment (hmlnarik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapStorageProviderFactory.java:146]: I think that instead of setting it, we should check the transaction level and log a warning if it is lower than that. Perhaps

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java (hunk 4 above)
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java:60]: NPE is comming from keycloak code, during startup there are default clients created and when shared instance is used none clients are persisted in database so when keycloak tries to use previously created client it gets `null`. I can try to investigate it further, but I don't exactly understand why we want to use shared instance in the first place. 

When we use shared instance, then we'd override values from instance by another instance and it doesn't seems correct. Could you please elaborate why there should be shared instance used?

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientModelCriteriaBuilder.java (hunk 1 above)
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientModelCriteriaBuilder.java:72]: Without it, it throws PSQLException:

```
Caused by: org.postgresql.util.PSQLException: ERROR: invalid input syntax for type json
  Detail: Token "c3a7d887" is invalid.
  Where: JSON data, line 1: c3a7d887...
```

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/delegate/JpaClientDelegateProvider.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                    case ATTRIBUTES:
+                        if (Hibernate.isPropertyInitialized(delegate, "attributes")) return delegate;
+
+                        CriteriaBuilder cb = em.getCriteriaBuilder();
+                        CriteriaQuery query = cb.createQuery(JpaClientEntity.class);
+                        Root root = query.from(JpaClientEntity.class);
+                        root.fetch("attributes", JoinType.INNER);
+                        query.select(root).where(cb.equal(root.get("id"), delegate.getId()));
+
+                        delegate = (JpaClientEntity) em.createQuery(query).getSingleResult();
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/delegate/JpaClientDelegateProvider.java]: no, it's bug :)

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java (hunk 4 above)
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java]: As discussed offline, instantiation of `JpaClientMetadata` was moved to constructor.

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java]: We need to keep `@Column` and `@Basic`, `@Generated` was removed

File: model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/liquibase/JpaClientLiquibaseUpdaterProviderFactory.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/liquibase/JpaClientLiquibaseUpdaterProviderFactory.java]: As discussed, there is new `MapJpaUpdaterSpi` introduced. It'll server for all map jpa implementations.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/JpaClientMapStorageProviderFactory.java [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+    private void customChanges(Connection connection, String schema, KeycloakSession session, MapJpaUpdaterProvider updater) {
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession lockSession) -> {
+            DBLockManager dbLockManager = new DBLockManager(lockSession);
+            DBLockProvider dbLock2 = dbLockManager.getDBLock();
+            dbLock2.waitForLock(DBLockProvider.Namespace.DATABASE);
+            try {
+                updater.update(ClientModel.class, connection, schema);
+            } finally {
+                dbLock2.releaseLock();
+            }
```
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/JpaClientMapStorageProviderFactory.java:187]: TODO not in this PR: The locking needs to be handled by the provider factory on the area level, i.e. here for client tables. There should be no separate locking mechanism necessary, all the locking has to be done in here, and the locked area should not prevent locking and updating other area, e.g. realms or users.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/JpaClientModelCriteriaBuilder.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+                            cb.function("->", JsonbType.class, root.get("metadata"), cb.literal("fScopeMappings")),
+                            cb.literal("\"" + value[0].toString().replace("\"", "\\\"") + "\"")))
+                    );
+                } else if (modelField.equals(SearchableFields.ALWAYS_DISPLAY_IN_CONSOLE)) {
+                    validateValue(value, 1, modelField, op);
+
+                    return new JpaClientModelCriteriaBuilder((cb, root) -> 
+                        cb.equal(
+                            cb.function("->", JsonbType.class, root.get("metadata"), cb.literal("fAlwaysDisplayInConsole")), 
+                            cb.literal(value[0]))
```
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/JpaClientModelCriteriaBuilder.java:80]: The `value[0]` can be misused for SQL injection.
Please introduce a method to validate the type and / or escape every value used in `literal` method.
Perhaps the `validateValue` could take one more parameter - the required types?

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/hibernate/jsonb/JsonbType.java [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+        private Object extractJson(CallableStatement statement, int index) throws SQLException {
+            return statement.getObject(index);
+        }
+
+        private Object extractJson(CallableStatement statement, String name) throws SQLException {
+            return statement.getObject(name);
+        }
+    }
+
+    private static class JsonbJavaTypeDescriptor extends AbstractTypeDescriptor<Object> {
```
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/hibernate/jsonb/JsonbType.java:151]: This is a very elegant way for no-downtime upgrade! 👍🏻

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/updater/liquibase/MapJpaLiquibaseUpdaterProvider.java [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+import org.keycloak.models.map.storage.jpa.updater.MapJpaUpdaterProvider.Status;
+
+public class MapJpaLiquibaseUpdaterProvider implements MapJpaUpdaterProvider {
+
+    private static final Logger logger = Logger.getLogger(MapJpaLiquibaseUpdaterProvider.class);
+
+    private static final Map<Class<?>, String> MODEL_TO_CHANGELOG_FILE = new HashMap<>();
+    static {
+        MODEL_TO_CHANGELOG_FILE.put(ClientModel.class, "META-INF/jpa-clients-changelog.xml");
+    }
```
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/updater/liquibase/MapJpaLiquibaseUpdaterProvider.java:52]: Can you leverage newly added https://github.com/keycloak/keycloak/blob/95614e8b4022306ffac328b85130403cc157fb18/model/map/src/main/java/org/keycloak/models/map/storage/ModelEntityUtil.java#L125 rather than introducing this map?

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/updater/liquibase/MapJpaLiquibaseUpdaterProviderFactory.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+import org.keycloak.Config;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.models.map.storage.jpa.updater.MapJpaUpdaterProvider;
+import org.keycloak.models.map.storage.jpa.updater.MapJpaUpdaterProviderFactory;
+
+public class MapJpaLiquibaseUpdaterProviderFactory implements MapJpaUpdaterProviderFactory {
+
+    public static final String PROVIDER_ID = "map-updater";
```
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/updater/liquibase/MapJpaLiquibaseUpdaterProviderFactory.java:28]:
```suggestion
    public static final String PROVIDER_ID = "map-liquibase-updater";
```

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thank you @vramik, looks very good! Could you please check few remaining inline comments?

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/JpaClientModelCriteriaBuilder.java [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+                        cb.equal(root.get(modelField.getName()), value[0])
+                    );
+                } else if (modelField.equals(SearchableFields.SCOPE_MAPPING_ROLE)) {
+                    validateValue(value, 1, modelField, op);
+
+                    return new JpaClientModelCriteriaBuilder((cb, root) -> 
+                        cb.isTrue(cb.function("@>",
+                            Boolean.TYPE,
+                            cb.function("->", JsonbType.class, root.get("metadata"), cb.literal("fScopeMappings")),
+                            cb.literal("\"" + value[0].toString().replace("\"", "\\\"") + "\"")))
```
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/JpaClientModelCriteriaBuilder.java:72]: This escaping should be perhaps extracted to a separate method used also from other `literal` invocations upon a JSON string.
Also I'm wondering if this is sufficient. Could we leverage Jackson, e.g.
```java
System.out.println(JsonbType.MAPPER.writeValueAsString("hello\"my friend,\nhow are'you?&amp;"));
```
outputs
```
"hello\"my friend,\nhow are'you?&amp;"
```

File: model/map-jpa/src/main/resources/META-INF/jpa-clients-changelog-1.xml [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
+
+    <!-- format of id of changeSet: clients-${JpaClientMapStorage.SUPPORTED_VERSION} -->
+    <changeSet author="keycloak" id="clients-1">
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map-jpa/src/main/resources/META-INF/jpa-clients-changelog-1.xml:23]: We should ideally be consistent across map storages, and there plural is used:
https://github.com/keycloak/keycloak/blob/654f1f74ff7684fb216893473ceac0930ed09bcb/model/map/src/main/java/org/keycloak/models/map/storage/ModelEntityUtil.java#L61-L80

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/updater/liquibase/MapJpaLiquibaseUpdaterProvider.java [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+        update(modelType, connection, file, defaultSchema);
+    }
+
+    private void update(Class modelType, Connection connection, File file, String defaultSchema) {
+        logger.debug("Starting database update");
+
+        // Need ThreadLocal as liquibase doesn't seem to have API to inject custom objects into tasks
+        ThreadLocalSessionContext.setCurrentSession(session);
+
+        Writer exportWriter = null;
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/updater/liquibase/MapJpaLiquibaseUpdaterProvider.java:76]: This functionality is copied from original updater. There it is used for manual migration, where migration sql script is generated instead of direct migration of database. I assumed we'd want to also have this possibility with new store implementations so I didn't remove it. Although I'm not sure if the assumption is correct :) wdyt?

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java [hunk 18]
```diff
@@ -0,0 +1,616 @@
...
+
+    @Override
+    public void removeClientScope(String id) {
+        checkEntityVersionForUpdate();
+        metadata.removeClientScope(id);
+    }
+
+    @Override
+    public MapProtocolMapperEntity getProtocolMapper(String id) {
+        return metadata.getProtocolMapper(id);
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java:194]: We need to check if metadata was initialized only for fields which might be loaded by `MapKeycloakTransaction.read(QueryParameters)` where there are currently loaded these fields: https://github.com/vramik/keycloak/blob/wip-JpaClientMapStorage/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/JpaClientMapKeycloakTransaction.java#L79-L83

When accessing other fields like here ProtocolMapper metadata (json) field will be loaded anyway.

File: model/map-jpa/src/main/resources/META-INF/jpa-clients-changelog-1.xml [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+                realmId varchar(36) generated always as (metadata->>'fRealmId') stored,
+                clientId varchar(255) generated always as (metadata->>'fClientId') stored,
+                protocol varchar(36) generated always as (metadata->>'fProtocol') stored,
+                enabled boolean generated always as ((metadata->>'fEnabled')::boolean) stored,
+                metadata jsonb
+            );
+
+            create index client_entityVersion on client(entityVersion);
+            create index client_realmId on client(realmId);
+            create index client_clientId on client(clientId);
```
Comment (hmlnarik) [inline model/map-jpa/src/main/resources/META-INF/jpa-clients-changelog-1.xml:38]:
```suggestion
            create index client_r_c on client(realmId, clientId);
```

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/JpaClientMapStorageProviderFactory.java (hunk 10 above)
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/JpaClientMapStorageProviderFactory.java:199]: https://issues.redhat.com/browse/KEYCLOAK-19896

**Review by sguilhen — DISMISSED** [review]

File: server-spi-private/src/main/java/org/keycloak/component/ComponentModelScope.java [hunk 20]
```diff
@@ -84,7 +84,8 @@ public Integer getInt(String key) {
     @Override
     public Integer getInt(String key, Integer defaultValue) {
         final String res = componentConfig.get(prefix + key, null);
-        return (res == null) ? origScope.getInt(key, defaultValue) : Integer.parseInt(res);
+        if (res == null) return origScope.getInt(key, defaultValue);
+        return Integer.parseInt(res);
```
Comment (hmlnarik) [inline server-spi-private/src/main/java/org/keycloak/component/ComponentModelScope.java:88]: Why is this change here?

Comment (vramik) [reply server-spi-private/src/main/java/org/keycloak/component/ComponentModelScope.java:88]: I am not sure why but without throws NPE:
```
Caused by: java.lang.NullPointerException
	at org.keycloak.component.ComponentModelScope.getInt(ComponentModelScope.java:87)
	at org.keycloak.component.ComponentModelScope.getInt(ComponentModelScope.java:81)
	at org.keycloak.models.map.storage.jpa.client.JpaClientMapStorageProviderFactory.lazyInit(JpaClientMapStorageProviderFactory.java:147)
```

Comment (vramik) [reply server-spi-private/src/main/java/org/keycloak/component/ComponentModelScope.java]: https://github.com/keycloak/keycloak/pull/9139

**Review by hmlnarik — APPROVED** [review]
Thank you @vramik. Great job!
